### ISSUE TITLE
Obtaining cycles from directed multi-graph

### ISSUE THREAD

**Issue by djelenc** [jgrapht/jgrapht#452]
The interface for finding cycles defines a cycle as a list of vertexes. However, if your graph is a directed multi-graph, this is a bit limiting. Consider the following graph and the corresponding code that was used to make it.

![f1](https://user-images.githubusercontent.com/2518358/33894120-7752bc30-df5d-11e7-868e-c39f3fbf62f4.png)

```java
static class LabeledEdge extends DefaultEdge {
    final String src;
    final String tgt;
    final String label;

    LabeledEdge(String src, String tgt, String label) {
        this.src = src;
        this.tgt = tgt;
        this.label = label;
    }

    public String toString() {
        return String.format("%s (%s-%s)", label, src, tgt);
    }

    @Override
    public boolean equals(Object o) {
        // omitted for brevity
    }

    @Override
    public int hashCode() {
        // omitted for brevity
    }
}

@Test
public void multiGraphCycles() throws ExportException, IOException {
    final Graph<String, LabeledEdge> graph = new DirectedMultigraph<>(LabeledEdge.class);

    graph.addVertex("A");
    graph.addVertex("B");
    graph.addVertex("C");
    graph.addEdge("A", "B", new LabeledEdge("A", "B", "a"));
    graph.addEdge("B", "A", new LabeledEdge("B", "A", "b"));
    graph.addEdge("B", "A", new LabeledEdge("B", "A", "c"));
    graph.addEdge("B", "C", new LabeledEdge("B", "C", "d"));
    graph.addEdge("C", "A", new LabeledEdge("C", "A", "e"));

    final DirectedSimpleCycles<String, LabeledEdge> algorithm = new HawickJamesSimpleCycles<>(graph);
    final List<List<String>> cycles = algorithm.findSimpleCycles();

    System.out.println(cycles);
}
```

Running the HawickJamesSimpleCycles---an algorithm that can work with multi-graphs---correctly returns three cycles [[B, A], [B, A], [C, B, A]]. 

However, this result only tells me that there are three cycles, but it does not give me enough information for traversal. The reason is that in a multi-graph, vertexes can be connected with multiple edges. The sequence of vertexes does not specify which edges form the cycle.

Additionally, it would be great if the returned sequence formed a complete cycle. While completing the cycle in a non-multi-graph is trivial (connect the last and the first vertex), this is not the case in a multi-graph: One has to know the entire path that forms the cycle.

I assume that changing the interface of `DirectedSimpleCycles.findSimpleCycles` to return a `List<GraphPath<V, E>>` instead of `List<List<V>>` and requiring all implementations the `findSimpleCycles` returns a connected `GraphPath` would solve the issue.

Any thoughts? What would be the best approach to this?

**Comment by jsichi** [comment]
Makes sense.  This would be a breaking interface change, so we can't do it in one step.  Instead, I would recommend deprecating the existing interfaces in package alg.cycle and defining the modified interface in package alg.interfaces.  This would bring alg.cycle into conformance with the pattern we've been using for the rest of the alg packages.

**Comment by eschava** [comment]
Also looks like HawickJamesSimpleCycles returns cycles in reverse order what has been fixed for JohnsonSimpleCycles in https://github.com/jgrapht/jgrapht/issues/337

**Comment by d-michail** [comment]
We have this identified as something that needs to be done in our [wiki](https://github.com/jgrapht/jgrapht/wiki/Google-Summer-Of-Code-2018#improve-the-cycles-support-of-jgrapht).

The task is to:
```
Add interface which returns cycles as list of edges or a `GraphPath` and change the 4 algorithms to follow the interface.
```
